libvast/src/system/type_registry.cpp

```diff
@@ -102,30 +117,52 @@ type_registry(type_registry_actor self, const path& dir) {
         self, "failed to persist state to disk:", self->system().render(err));
     self->quit(msg.reason);
   });
-  return {[=](telemetry_atom) {
-            // Send out a heartbeat.
-            self->send(self->state.accountant, self->state.telemetry());
-            self->delayed_send(self, defaults::system::telemetry_rate,
-                               telemetry_atom::value);
-          },
-          [=](status_atom) {
-            // Send out a status report.
-            return self->state.status();
-          },
-          [=](caf::stream<table_slice_ptr> in) {
-            // Store layout of every incoming table slice.
-            caf::attach_stream_sink(
-              self, in,
-              [=](caf::unit_t&) { VAST_DEBUG(self, "initialized stream"); },
-              [=](caf::unit_t&, table_slice_ptr x) {
-                VAST_TRACE(self, "received new table slice");
-                self->state.insert(std::move(x->layout()));
-              });
-          },
-          [=](std::string name) {
-            // Retrieve a list of known states for a name.
-            return self->state.lookup(name);
-          }};
+  // Load existing state from disk if possible.
+  if (auto err = self->state.load_from_disk())
+    self->quit(std::move(err));
+  // Load already known types from the event_types singleton.
+  auto schema = vast::event_types::get();
+  VAST_ASSERT(schema);
+  for (auto& type : *schema)
+    self->send(self, put_atom::value, type);
+  // The behavior of the type-registry.
+  return {
+    [=](telemetry_atom) {
+      VAST_TRACE(self, "sends out a telemetry report to the",
+                 VAST_ARG("accountant", self->state.accountant));
+      self->send(self->state.accountant, self->state.telemetry());
+      self->delayed_send(self, defaults::system::telemetry_rate,
+                         telemetry_atom::value);
+    },
+    [=](status_atom) {
+      VAST_TRACE(self, "sends out a status report");
+      return self->state.status();
+    },
+    [=](caf::stream<table_slice_ptr> in) {
+      VAST_TRACE(self, "attaches to", VAST_ARG("stream", in));
+      caf::attach_stream_sink(
+        self, in,
+        [=](caf::unit_t&) {
+          // nop
+        },
+        [=](caf::unit_t&, table_slice_ptr x) {
+          self->state.insert(std::move(x->layout()));
```

Comment:
I know this was there before: but why are we *moving* from an immutable object?

Reply:
I'll create a follow-up story for this.